axlearn/common/flash_attention/gpu_attention_test.py

```diff
@@ -174,56 +176,65 @@ def ref_fn(q, k, v, bias):
     chex.assert_trees_all_close(jax_grads, jax_ref_grads, atol=0.05)
 
 
-# We also include a test for Triton with Pallas, to cross validate the triton
-# compatibility with our own implementation.
+# We test the cudnn_dot_product_attention against the reference flash_attention.
+# Due to its algorithmic equivalence, the outputs should be close in both fp16 and bfloat16.
 @pytest.mark.parametrize(
     "batch_size,num_heads,seq_len,per_head_dim",
     [
-        (2, 2, 384, 64),
+        (1, 2, 2048, 128),
+        (2, 2, 2048, 128),
+        (1, 4, 4096, 128),
+        (2, 8, 4096, 128),
     ],
 )
 @pytest.mark.parametrize("causal", [True, False])
-@pytest.mark.parametrize("bias_type", ["none", "vector"])
+@pytest.mark.parametrize("dtype", [jnp.bfloat16, jnp.float16])
 @pytest.mark.skipif(jax.devices()[0].platform != "gpu", reason="Test only runs on GPU.")
-def test_mha_against_pallas_ref(
+def test_cudnn_against_triton_ref(
     batch_size: int,
     num_heads: int,
     seq_len: int,
     per_head_dim: int,
     causal: bool,
-    bias_type: str,
+    dtype: jnp.dtype,
 ):
     q = jax.random.normal(
-        jax.random.PRNGKey(0), (batch_size, seq_len, num_heads, per_head_dim), dtype=jnp.float16
+        jax.random.PRNGKey(0), (batch_size, seq_len, num_heads, per_head_dim), dtype=dtype
     )
     k = jax.random.normal(
-        jax.random.PRNGKey(1), (batch_size, seq_len, num_heads, per_head_dim), dtype=jnp.float16
+        jax.random.PRNGKey(1), (batch_size, seq_len, num_heads, per_head_dim), dtype=dtype
     )
     v = jax.random.normal(
-        jax.random.PRNGKey(2), (batch_size, seq_len, num_heads, per_head_dim), dtype=jnp.float16
+        jax.random.PRNGKey(2), (batch_size, seq_len, num_heads, per_head_dim), dtype=dtype
     )
     # Make sure that it is running on GPU.
     assert str(q.devices()) == "{cuda(id=0)}"
 
     sm_scale = q.shape[-1] ** -0.5
-    if bias_type == "vector":
-        segment_left = jnp.ones((batch_size, seq_len // 2), dtype=jnp.int32)
-        segment_right = jnp.zeros((batch_size, seq_len // 2), dtype=jnp.int32)
-        segment_ids = jnp.concatenate([segment_left, segment_right], axis=-1)
-    else:
-        segment_ids = None
+
     # Compare outputs.
-    jax_out = mha_reference(q, k, v, bias=segment_ids, causal=causal, softmax_scale=sm_scale)
-    jax_ref_out = pallas_mha(q, k, v, segment_ids=segment_ids, causal=causal, sm_scale=sm_scale)
-    chex.assert_trees_all_close(jax_out, jax_ref_out, atol=0.005, rtol=1e-5)
+    jax_out = cudnn_dot_product_attention(q, k, v, bias=None, causal=causal, softmax_scale=sm_scale)
+    jax_ref_out = flash_attention(q, k, v, bias=None, causal=causal, softmax_scale=sm_scale)
+    if dtype == jnp.bfloat16:
+        # We relax the atol to support bf16 in the unit test.
+        chex.assert_trees_all_close(jax_out, jax_ref_out, atol=0.02, rtol=1e-5)
+    elif dtype == jnp.float16:
+        chex.assert_trees_all_close(jax_out, jax_ref_out, atol=0.005, rtol=1e-5)
```

Comment:
We should probably add an `else: raise` or similar. If someone parameterizes a different dtype, we don't want to inadvertently skip the check.

Reply:
done